libparsec/crates/types/src/local_manifest.rs

```diff
@@ -857,6 +888,15 @@ impl LocalFolderManifest {
         )
     }
 
+    /// Apply the prevent sync pattern *on the local children* (i.e. not on
+    /// `base.children` as one could expect) to remove from it all the confined
+    /// entries and collect them into `remote_confinement_points`.
+    ///
+    /// This method is expected to be run on a local manifest that have just been created
+    /// from a remote manifest. This way it local children are currently a mere copy
+    /// of it remote children (i.e. `base.children`) and it's this method's goal to
+    /// filter out the entries that should be remotely confined (i.e. entries that
+    /// should be kept in `base.children`, but not appear in local children).
```

Comment:
That's not exactly true: it's also called when applying a new pattern, after restoring the remote confinement points.

But now that I checked the implementation, I think there's an issue: the new remote confinement points should be the entries that appear both in `children` AND `base.children` and that have a remote name that matches the `prevent_sync_pattern`.

So I would do 3 things:
- fix the implementation (and see how it reflects on the tests)
- update the docstring to explain that this method filters out entries from `children` that have a corresponding remote name that matches the `prevent_sync_pattern`
- update the docstring to note that this method is expected to run on a local manifest that doesn't have remote confinement point (i.e either a local manifest that has just been created from a remote manifest, or a local manifest that had their remote confinement points restored).

Reply:
I've updated the comment:
```rust
    /// Apply the prevent sync pattern *on the local children* (i.e. not on
    /// `base.children` as one could expect) to remove from it all the confined
    /// entries and collect them into `remote_confinement_points`.
    ///
    /// This method is expected to be run on a local manifest where it's local children
    /// are currently a mere copy of it remote children (i.e. `base.children`).
    /// This occurs in two places:
    /// - When creating a new local manifest from a remote manifest.
    /// - When applying a new prevent sync pattern, in which case `filter_local_confinement_points`
    ///   and `restore_remote_confinement_points` has just been previously called to revert
    ///   the local children according to the local&remote confinement points.
    ///
    /// Once in this state, it is the method's goal to filter out the entries that
    /// should be remotely confined (i.e. entries that should be kept in `base.children`,
    /// but not appear in local children).
```

concerning the change in the implementation, I'm totally puzzled on what is the reasoning here...

I've considered that when this method is called we should have `self.children == self.base.children`, however you say we should check for both `self.children` and `self.base.children` which would indicate my assumption is wrong.
Can you give an example where calling `filter_local_confinement_points` then `restore_remote_confinement_points` doesn't lead to a local children equal to it remote (I suspect this has to do with renaming fields) ? I think it would be good to explain this case in the documentation.

Reply:
> Can you give an example

Imagine this local file manifest with an old `prevent_sync_pattern` that matches nothing:
```
LocalFolderManifest
- base.children: {}
- children: {"a.tmp": 1}
- local_confinement_points: []
- remote_confinement_points: []
```

Note that a `a.tmp` has been created locally but it hasn't been synced yet.

Then we call `apply_prevent_sync_pattern` on it with the `*.tmp` filter. Remember this implementation:
```rust
        let result = self.clone();
        result
            .filter_local_confinement_points()
            .restore_remote_confinement_points()
            .filter_remote_entries(prevent_sync_pattern)
            .restore_local_confinement_points(self, prevent_sync_pattern, timestamp)
```

So after `filter_local_confinement_points` and `restore_remote_confinement_points` the manifest hasn't changed (cause it has no local nor remote confinement points). But after `filter_remote_entries(prevent_sync_pattern)`, with the current implementation, the manifest becomes:
```
LocalFolderManifest
- base.children: {}
- children: {}
- local_confinement_points: []
- remote_confinement_points: [1]
```
This breaks my assumption that all entries in `remote_confinement_points` should appear in `base.children`. This is surprising but doesn't lead to a bug, due to the implementation of `restore_local_confinement_points` which will restore `a.tmp` due to `1` being in `self.remote_confinement_points`. The end manifest is:
```
LocalFolderManifest
- base.children: {}
- children: {"a.tmp": 1}
- local_confinement_points: [1]
- remote_confinement_points: [1]
```
One might think that `1` being part of the `remote_confinement_points` is an issue, but it will be simply cleared by the next call to `restore_remote_confinement_points` (because the logic is: restore any child present in `base.children` if its id is present in `remote_confinement_points`, then clear `remote_confinement_points`)